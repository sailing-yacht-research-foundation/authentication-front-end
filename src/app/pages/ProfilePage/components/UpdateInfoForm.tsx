import 'react-phone-number-input/style.css';

import React, { useState } from 'react';
import { Input, Form, Button, Divider, DatePicker, Spin, Space } from 'antd';
import PhoneInput from 'react-phone-number-input';
import moment from 'moment';
import styled from 'styled-components';
import { getUserAttribute } from 'utils/user-utils';
import Auth from '@aws-amplify/auth';
import { toast } from 'react-toastify';

const layout = {
    labelCol: { xs: { span: 24 }, sm: { span: 24 }, lg: { span: 8 } },
    wrapperCol: { xs: { span: 24 }, sm: { span: 24 }, md: { span: 8 }, lg: { span: 8 } },
};

const tailLayout = {
    wrapperCol: { xs: { span: 24 }, sm: { span: 24 }, md: { span: 12, offset: 8 }, lg: { span: 12, offset: 8 } }
};

const format = "DD.MM.YYYY HH:mm";

const disabledDates = [
    {
        start: "29.12.2002 13:00",
        end: moment().format(format)
    },
];

export const UpdateInfo = (props) => {
    const [isUpdatingProfile, setIsUpdatingProfile] = useState<boolean>(false);

    const { authUser } = props;

    const onFinish = (values) => {
        const { name, phone_number, sailing_number, address, facebook, instagram, twitter, birthdate } = values;

        setIsUpdatingProfile(true);

        Auth.currentAuthenticatedUser().then(user => {
            setIsUpdatingProfile(false);
            Auth.updateUserAttributes(user, {
                name: name,
                phone_number: phone_number,
                address: address,
                birthdate: birthdate ? birthdate.format("YYYY-MM-DD") : moment('2002-01-01').format("YYYY-MM-DD"),
                'custom:sailing_number': sailing_number,
                'custom:facebook': facebook,
                'custom:instagram': instagram,
                'custom:twitter': twitter,
            }).then(response => {
                toast.success('Your profile has been successfully updated!');
                props.cancelUpdateProfile();
            }).catch(error => {
                console.log(error);
            })
        }).catch(error => {
            toast.error(error.message);
            setIsUpdatingProfile(false);
        })
    }

    return (
        <Wrapper>
            <Spin spinning={isUpdatingProfile} tip="Updating your profile...">
                <Form
                    {...layout}
                    name="basic"
                    initialValues={{
                        email: getUserAttribute(authUser, 'email'),
                        name: getUserAttribute(authUser, 'name'),
                        phone_number: getUserAttribute(authUser, 'phone_number'),
                        address: getUserAttribute(authUser, 'address'),
                        birthdate: moment(getUserAttribute(authUser, 'birthdate')),
                        sailing_number: getUserAttribute(authUser, 'custom:sailing_number'),
                        facebook: getUserAttribute(authUser, 'custom:facebook'),
                        instagram: getUserAttribute(authUser, 'custom:instagram'),
                        twitter: getUserAttribute(authUser, 'custom:twitter'),
                    }}
                    onFinish={onFinish}
                >
                    <Form.Item
                        label="Email"
                        name="email"
                        rules={[{ required: true, type: 'email' }]}
                    >
                        <Input disabled />
                    </Form.Item>

                    <Form.Item
                        label="Name"
                        name="name"
                        rules={[{ required: true }]}
                    >
                        <Input />
                    </Form.Item>

                    <Divider />

                    <Form.Item
                        label="Phone Number"
                        name="phone_number"
                        rules={[{ type: 'string' }]}
                    >
                        <PhoneInput className="ant-input"
                            placeholder="Enter phone number" />
                    </Form.Item>

                    <Form.Item
                        label="Address"
                        name="address"
                    >
                        <Input />
                    </Form.Item>

                    <Form.Item
                        label="Date Of Birth"
                        name="birthdate"
                        rules={[{ type: 'date' }]}
                    >
                        <DatePicker
                            ref="datePickerRef"
                            style={{ width: '100%' }}
                            defaultValue={moment('2002-01-01')}
                            showToday={false}
                            disabledDate={current => {
                                return disabledDates.some(date =>
                                    current.isBetween(
                                        moment(date["start"], format),
                                        moment(date["end"], format)
                                    )
                                );
                            }}
                            dateRender={current => {
                                return (
                                    <div className="ant-picker-cell-inner">
                                        {current.date()}
                                    </div>
                                );
                            }}
                        />
                    </Form.Item>

                    <Form.Item
                        label="World Sailing Number"
                        name="sailing_number"
                    >
                        <Input />
                    </Form.Item>

                    <Form.Item
                        label="Facebook profile"
                        name="facebook"
                        rules={[{ type: 'url' }]}
                    >
                        <Input />
                    </Form.Item>

                    <Form.Item
                        label="Instagram profile"
                        name="instagram"
                        rules={[{ type: 'url' }]}
                    >
                        <Input />
                    </Form.Item>

                    <Form.Item
                        label="Twitter profile"
                        name="twitter"
                        rules={[{ type: 'url' }]}
                    >
                        <Input />
                    </Form.Item>

                    <Divider />

                    <Form.Item {...tailLayout}>
                        <Space size={15}>
                            <Button type="primary" htmlType="submit">
                                Update
                            </Button>

                            <Button htmlType="button" onClick={(e) => {
                                e.preventDefault();
                                props.cancelUpdateProfile();
                            }}>
                                Cancel
                            </Button>
                        </Space>
                    </Form.Item>

                </Form>
            </Spin>
        </Wrapper>
    )
}

const Wrapper = styled.div`
margin-top: 50px;
width: 100%;
`