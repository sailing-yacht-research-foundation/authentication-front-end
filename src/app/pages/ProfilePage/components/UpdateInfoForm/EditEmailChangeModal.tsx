import React from 'react';
import { Modal } from 'antd';
import styled from 'styled-components';
import { getUserAttribute } from 'utils/user-utils';
import { useTranslation } from 'react-i18next';
import { translations } from 'locales/translations';

export const EditEmailChangeModal = (props) => {

    const { t } = useTranslation();

    const { showEmailChangeAlertModal,
        formFieldsBeforeUpdate,
        defaultFormFields,
        authUser,
        form,
        updateUserInfo,
        setShowEmailChangeAlertModal,
        setFormFieldsBeforeUpdate } = props;

    return (
        <Modal title={t(translations.profile_page.update_profile.email_change_confirmation)}
            visible={showEmailChangeAlertModal}
            onOk={() => {
                updateUserInfo(formFieldsBeforeUpdate);
                setShowEmailChangeAlertModal(false);
            }}
            onCancel={() => {
                setShowEmailChangeAlertModal(false);
                setFormFieldsBeforeUpdate(defaultFormFields);
                form.setFieldsValue({ // reset the email to the last state.
                    email: getUserAttribute(authUser, 'email')
                });
            }}>
            <EmailChangeMessageText>
                {t(translations.profile_page.update_profile.hey_your_about_to_change_your_email, { name: getUserAttribute(authUser, 'name') })}
            </EmailChangeMessageText>
        </Modal>
    )
}

const EmailChangeMessageText = styled.div`
margin-bottom: 10px;
color: #f81d22;
`;