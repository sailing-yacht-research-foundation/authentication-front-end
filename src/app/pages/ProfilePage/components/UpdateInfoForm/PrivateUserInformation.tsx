import React, { useState, useEffect, useRef } from 'react';
import { Select, Col, Row, Form, DatePicker, Menu } from 'antd';
import PlacesAutocomplete from 'react-places-autocomplete';
import moment from 'moment';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

import {
    SyrfFieldLabel,
    SyrfInputField,
    SyrfPhoneInput,
    SyrfFormTitle,
    SyrfFormSelect,
} from 'app/components/SyrfForm';
import { languagesList } from 'utils/languages-util';
import { translations } from 'locales/translations';
import { StyleConstants } from 'styles/StyleConstants';
import { checkForVerifiedField, getUserAttribute } from 'utils/user-utils';
import { FIELD_VALIDATE } from 'utils/constants';
import ReactTooltip from 'react-tooltip';
import { FilterWorldSailingNumber } from 'utils/world-sailing-number';
import countryCodeSource from '../../assets/world-sailing-number-countrycode.json';

const format = "DD.MM.YYYY HH:mm";

const disabledDates = [
    {
        start: "29.12.2002 13:00",
        end: moment().format(format)
    },
];

export const PrivateUserInformation = (props) => {

    const { authUser, address, setAddress } = props;

    const { t } = useTranslation();
    const [isSuggestionVisible, setIsSuggestionVisible] = useState(false);
    const [worldSailingNumber, setWorldSailingNumber] = useState('');
    const [countryCodeList, setCountryCodeList] = useState<string[]>([]);

    const isInitRef = useRef(true);

    useEffect(() => {
        if (worldSailingNumber.length > 2) setIsSuggestionVisible(false);
        else {
            if (!isInitRef.current) {
                setIsSuggestionVisible(true);
            };

            handleFilterWorldSailingNumber(worldSailingNumber, countryCodeSource);
        }

        isInitRef.current = false;
    }, [worldSailingNumber])

    const renderLanguegesDropdownList = () => {
        const objectArray = Object.entries(languagesList);

        return objectArray.map(([key, value]) => {
            return <Select.Option key={key} value={key.toLowerCase()}>{value.nativeName}</Select.Option>
        });
    }

    const renderVerifiedStatus = (type: string) => {
        const verified = checkForVerifiedField(authUser, type);
        const userPhoneNumberExists = !!getUserAttribute(authUser, 'phone_number');

        if (type === FIELD_VALIDATE.email)
            return <ItemVerifyMessage className={verified ? 'verified' : ''}>{t(translations.profile_page.update_profile.your_email_is, { verify_status: (verified ? 'verified' : 'not verified') })}</ItemVerifyMessage>;

        if (userPhoneNumberExists) {
            return verified ? (<ItemVerifyMessage className={'verified'}>{t(translations.profile_page.update_profile.your_phone_is_verified)}</ItemVerifyMessage>) :
                (
                    <ItemVerifyMessage>{t(translations.profile_page.update_profile.your_phone_is_not_verified)}</ItemVerifyMessage>
                )
        }
    }

    const handleSuggestionVisible = () => {
        if (worldSailingNumber.length > 2) {
            setIsSuggestionVisible(false);
            return;
        };

        setIsSuggestionVisible(true);
    };

    const handleSuggestionNotVisible = () => {
        setIsSuggestionVisible(false);
    };

    const handleSuggestionBlur = () => {
        setTimeout(() => {
            handleSuggestionNotVisible();
        }, 200);
    };

    const handleWorldSailingNumberChange = (e) => {
        setWorldSailingNumber(e.target.value);
    };

    const handleFilterWorldSailingNumber = (key: string, source: string[]) => {
        const filtered = FilterWorldSailingNumber(key, source);
        setCountryCodeList(filtered);
    };

    const handleCountryCodeSuggestionClick = (countryCode) => {
        setWorldSailingNumber(countryCode);
    };

    const renderCountryCodeList = (countryCodeList: string[]) => {
        return countryCodeList.map((countryCode) => {
            return <SailingNumberSuggestionItem onClick={() => {handleCountryCodeSuggestionClick(countryCode)}} key={countryCode}>{countryCode}</SailingNumberSuggestionItem>
        });
    };



    return (
        <Wrapper>
            <SyrfFormTitle>{t(translations.profile_page.update_profile.private_user_details)}</SyrfFormTitle>
            <Form.Item
                label={<SyrfFieldLabel>Email</SyrfFieldLabel>}
                name="email"
                rules={[{ required: true, type: 'email' }]}
                data-tip={t(translations.tip.email)}
            >
                <SyrfInputField disabled />
            </Form.Item>
            {renderVerifiedStatus(FIELD_VALIDATE.email)}

            <Form.Item
                label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.address)}</SyrfFieldLabel>}
                name="address"
                data-tip={t(translations.tip.address)}
            >
                <PlacesAutocomplete
                    value={address}
                    onChange={(address) => { setAddress(address) }}
                >
                    {({ getInputProps, suggestions, getSuggestionItemProps }) => {
                        return (
                        <>
                            <SyrfInputField
                                {...getInputProps({
                                    placeholder: t(translations.profile_page.update_profile.search_places),
                                    className: 'location-search-input',
                                })}
                                value={address}
                                autoCorrect="off"
                            />
                            {suggestions.length > 0 && <StyledPLaceDropdown>
                                {suggestions.map((suggestion) => {
                                    const className = suggestion.active
                                        ? 'suggestion-item--active'
                                        : 'suggestion-item';
                                    // inline style for demonstration purpose
                                    const style = suggestion.active
                                        ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                                        : { backgroundColor: '#ffffff', cursor: 'pointer' };
                                    return (
                                        <Menu.Item
                                            {...getSuggestionItemProps(suggestion, {
                                                className,
                                                style,
                                            })}
                                            key={suggestion.index}
                                        >
                                            <span>{suggestion.description}</span>
                                        </Menu.Item>
                                    );
                                })}
                            </StyledPLaceDropdown>}
                        </>
                    )}}
                </PlacesAutocomplete>
            </Form.Item>

            <Row gutter={24}>
                <Col xs={24} sm={24} md={12} lg={12}>
                    <Form.Item
                        label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.date_of_birth)}</SyrfFieldLabel>}
                        name="birthdate"
                        data-tip={t(translations.tip.date_of_birth)}
                        rules={[{ type: 'date' }, {
                            required: true,
                            message: t(translations.forms.birth_date_is_required)
                        }]}
                    >
                        <DatePicker
                            style={{ width: '100%' }}
                            className="syrf-datepicker"
                            showToday={false}
                            disabledDate={current => {
                                return disabledDates.some(date =>
                                    current.isBetween(
                                        moment(date["start"], format),
                                        moment(date["end"], format)
                                    )
                                );
                            }}
                            dateRender={current => {
                                return (
                                    <div className="ant-picker-cell-inner">
                                        {current.date()}
                                    </div>
                                );
                            }}
                        />
                    </Form.Item>
                </Col>

                <Col xs={24} sm={24} md={12} lg={12}>
                    <Form.Item
                        label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.world_sailing_number)}</SyrfFieldLabel>}
                        name="sailing_number"
                        data-tip={t(translations.tip.world_sailing_number)}
                        style={{ position: 'relative' }}
                    >
                        <SyrfInputField value={worldSailingNumber} onChange={handleWorldSailingNumberChange} autoCorrect="off" onFocus={handleSuggestionVisible} onBlur={handleSuggestionBlur} />
                        <SailingNumberSuggestionContainer style={{ display: isSuggestionVisible ? 'block' : 'none' }}>
                            <div style={{ overflowY: 'auto', maxHeight: '200px' }}>
                                {countryCodeList.length > 0 ?
                                    renderCountryCodeList(countryCodeList) :
                                    <SailingNumberSuggestionEmpty>No data found</SailingNumberSuggestionEmpty>
                                }
                            </div>
                        </SailingNumberSuggestionContainer>
                    </Form.Item>
                </Col>
            </Row>

            <Row gutter={24}>
                <Col xs={24} sm={24} md={12} lg={12}>
                    <Form.Item
                        label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.phone_number)}</SyrfFieldLabel>}
                        name="phone_number"
                        data-tip={t(translations.tip.phone_number)}
                        rules={[{ type: 'string' }]}
                    >
                        <SyrfPhoneInput
                            inputClass="syrf-phone-number-input"
                            buttonClass="syrf-flag-dropdown"
                            placeholder={t(translations.profile_page.update_profile.enter_phone_number)} />
                    </Form.Item>
                    {renderVerifiedStatus(FIELD_VALIDATE.phone)}
                </Col>

                <Col xs={24} sm={24} md={12} lg={12}>
                    <Form.Item
                        label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.language)}</SyrfFieldLabel>}
                        name="language"
                        data-tip={t(translations.tip.language)}
                        rules={[{ required: true }]}
                    >
                        <SyrfFormSelect placeholder={t(translations.profile_page.update_profile.select_a_language)}
                            showSearch
                            filterOption={(input, option) => {
                                if (option) {
                                    return option.props.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
                                        || option.props.value.toLowerCase().indexOf(input.toLowerCase()) >= 0
                                }

                                return false;
                            }}
                        >
                            {
                                renderLanguegesDropdownList()
                            }
                        </SyrfFormSelect>
                    </Form.Item>
                </Col>
            </Row>
            <ReactTooltip/>
        </Wrapper>
    )
}

const Wrapper = styled.div`
    background: #fff;
    padding: 30px 25px;
    border-radius: 10px;
    margin: 30px 0;
`;

const ItemVerifyMessage = styled.div`
    color: rgb(115, 116, 117);
    font-size: 13px;
    margin-top: -20px;
    text-align: right;

    &.verified {
        color: ${StyleConstants.MAIN_TONE_COLOR} !important;
    }
`;

const StyledPLaceDropdown = styled(Menu)`
    position: absolute;
    z-index: 2;
    background: #fff;
    border: 1px solid #d9d9d9;
    width: 100%;
`;

const SailingNumberSuggestionContainer = styled.div`
    position: absolute;
    top: 100%;
    padding: 4px 0px;
    border-radius: 4px;
    left: 0px;
    width: 100%;
    box-shadow: rgba(149, 157, 165, 0.2) 0px 8px 24px;
    z-index: 20;
    transform: translateY(8px);
`;

const SailingNumberSuggestionItem = styled.div`
    padding: 4px 16px;
    background-color: #FFFFFF;
    z-index: 20;

    &:hover {
        background-color: #E6F7FF;
    }
`;

const SailingNumberSuggestionEmpty = styled.div`
    padding: 4px 16px;
    background-color: #FFFFFF;
    z-index: 20;
`;