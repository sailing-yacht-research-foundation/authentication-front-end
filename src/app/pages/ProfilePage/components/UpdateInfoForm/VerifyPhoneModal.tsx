import React from 'react';
import { Modal, Form } from 'antd';
import {
    SyrfFieldLabel,
    SyrfInputField,
} from 'app/components/SyrfForm';
import { Auth } from 'aws-amplify';
import { toast } from 'react-toastify';
import styled from 'styled-components';
import { translations } from 'locales/translations';
import { useTranslation } from 'react-i18next';
import { showToastMessageOnRequestError } from 'utils/helpers';
import { verifyPhoneNumber } from 'services/live-data-server/user';
import { useDispatch } from 'react-redux';
import { UseLoginSlice } from 'app/pages/LoginPage/slice';

export const VerifyPhoneModal = (props) => {

    const { t } = useTranslation();

    const dispatch = useDispatch();

    const { actions } = UseLoginSlice();

    const {
        showPhoneVerifyModal,
        setShowPhoneVerifyModal,
        sendPhoneVerification,
    } = props;

    const [verifyPhoneForm] = Form.useForm();

    const verifyFormAndPhoneNumber = () => {
        verifyPhoneForm
            .validateFields()
            .then(async values => {
                const { code } = values;
                const response = await verifyPhoneNumber(code);
                if (response.success) {
                    toast.success(t(translations.profile_page.update_profile.your_phone_number_has_been_verified));
                    dispatch(actions.getUser());
                    setShowPhoneVerifyModal(false);
                } else {
                    showToastMessageOnRequestError(response.error);
                }
                verifyPhoneForm.resetFields();
            });
    }

    return (
        <Modal
            title={t(translations.profile_page.update_profile.please_verify_your_phone_number)}
            visible={showPhoneVerifyModal}
            onOk={() => {
                verifyFormAndPhoneNumber();
            }}
            onCancel={() => {
                setShowPhoneVerifyModal(false);
            }}>
            <Form
                form={verifyPhoneForm}
                layout="vertical"
                name="basic"
                initialValues={{
                    code: '',
                }}
            >
                <Form.Item
                    label={<SyrfFieldLabel>{t(translations.profile_page.update_profile.code)}</SyrfFieldLabel>}
                    name="code"
                    rules={[{ required: true, min: 6, message: t(translations.profile_page.update_profile.code_is_required) }]}
                >
                    <SyrfInputField
                        placeholder={t(translations.profile_page.update_profile.enter_the_code_you_received)}
                        type="number"
                    />
                </Form.Item>
            </Form>

            <div style={{ marginTop: '10px', textAlign: 'right' }}>
                <span>{t(translations.profile_page.update_profile.could_not_receive_the_code)} &nbsp; <a style={{ float: 'right' }} href="/" onClick={(e) => {
                    e.preventDefault();
                    sendPhoneVerification();
                }}>{t(translations.profile_page.update_profile.resend)}</a></span>
            </div>
        </Modal>
    )
}

const ModalMessage = styled.div`
    margin: 0 5px;
    margin-bottom: 15px;
`;