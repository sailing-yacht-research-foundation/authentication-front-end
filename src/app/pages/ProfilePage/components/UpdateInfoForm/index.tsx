import 'react-phone-input-2/lib/style.css';

import React, { useState } from 'react';
import { Form, Spin } from 'antd';
import moment from 'moment';
import styled from 'styled-components';
import { checkForVerifiedField, getUserAttribute } from 'utils/user-utils';
import Auth from '@aws-amplify/auth';
import { SyrfFormButton, SyrfFormWrapper } from 'app/components/SyrfForm';
import { FIELD_VALIDATE, removePlusFromPhoneNumber, replaceObjectPropertiesFromNullToEmptyString } from 'utils/helper';
import { PrivateUserInformation } from './PrivateUserInformation';
import { toast } from 'react-toastify';
import { PublicUserInformation } from './PublicUserInformation';
import { VerifyPhoneModal } from './VerifyPhoneModal';
import { EditEmailChangeModal } from './EditEmailChangeModal';
import { media } from 'styles/media';

const defaultFormFields = {
    email: '',
    name: '',
    phone_number: '',
    sailing_number: '',
    birthdate: '',
    language: '',
    country: '',
    share_social: ''
};

export const UpdateInfo = (props) => {
    const [isUpdatingProfile, setIsUpdatingProfile] = useState<boolean>(false);

    const { authUser } = props;

    const [showPhoneVerifyModal, setShowPhoneVerifyModal] = useState<boolean>(false);

    const [address, setAddress] = React.useState<string>(getUserAttribute(authUser, 'address') || '');

    const [phoneVerifyModalMessage, setPhoneVerifyModalMessage] = React.useState<string>('');

    const [showEmailChangeAlertModal, setShowEmailChangeAlertModal] = React.useState<boolean>(false);

    const [form] = Form.useForm();

    const [formHasBeenChanged, setFormHasBeenChanged] = React.useState<boolean>(false);

    // for storing fields filled by user from the form for later use when
    // user still chooses to change the email.
    const [formFieldsBeforeUpdate, setFormFieldsBeforeUpdate] = React.useState(defaultFormFields);

    const onFinish = (values) => {
        values = replaceObjectPropertiesFromNullToEmptyString(values);
        const { email } = values;

        if (checkForEmailChanged(email)) {
            setShowEmailChangeAlertModal(true);
            setFormFieldsBeforeUpdate(values);
        } else {
            updateUserInfo(values);
        }
    }

    const updateUserInfo = ({
        email,
        phone_number,
        sailing_number,
        birthdate,
        language,
        country,
        share_social,
        bio,
        first_name,
        last_name,
    }) => {

        setIsUpdatingProfile(true);

        Auth.currentAuthenticatedUser().then(user => {
            Auth.updateUserAttributes(user, {
                email: email,
                phone_number: phone_number ? removePlusFromPhoneNumber(phone_number) : '',
                address: address,
                birthdate: birthdate ? birthdate.format("YYYY-MM-DD") : moment('2002-01-01').format("YYYY-MM-DD"),
                locale: country,
                'custom:sailing_number': sailing_number,
                'custom:language': language,
                'custom:share_social': share_social ? 'y' : '', // y stands for yes
                'custom:first_name': first_name,
                'custom:last_name': last_name,
                'custom:bio': bio,
                name: first_name + ' ' + last_name
            }).then(response => {
                onUpdateProfileSuccess();
            }).catch(error => {
                toast.error(error.message);
                setIsUpdatingProfile(false);
                setFormFieldsBeforeUpdate(defaultFormFields);
            })
        }).catch(error => {
            toast.error(error.message);
            setIsUpdatingProfile(false);
            setFormFieldsBeforeUpdate(defaultFormFields);
        })
    }

    const onUpdateProfileSuccess = () => {
        setIsUpdatingProfile(false);
        toast.success('Your profile has been successfully updated!');
        props.cancelUpdateProfile();
        checkPhoneVerifyStatus();
        setFormFieldsBeforeUpdate(defaultFormFields);
        setFormHasBeenChanged(false);
    }

    const checkPhoneVerifyStatus = () => {
        Auth.currentAuthenticatedUser().then(user => {
            if (!!getUserAttribute(user, 'phone_number') && !checkForVerifiedField(user, FIELD_VALIDATE.phone)) { // user inputed phone and it's not verified
                showPhoneVerifyModalWithMessage(
                    `Hey ${getUserAttribute(user, 'name')}, your phone number is not verified, you will receive a call to verify your phone number.`
                );
            }
        }).catch((error) => {
            toast.error(error.message);
        })
    }

    const checkForEmailChanged = (email) => {
        const oldEmail = getUserAttribute(authUser, 'email');
        const newEmail = email;

        return oldEmail !== newEmail;
    }

    const showPhoneVerifyModalWithMessage = (message: string) => {
        setPhoneVerifyModalMessage(message);
        setShowPhoneVerifyModal(true);
        sendPhoneVerification();
    }

    const sendPhoneVerification = () => {
        Auth.verifyCurrentUserAttribute('phone_number').then(() => {
            toast.success('You will receive a phone call to verify your phone number.');
        }).catch(error => {
            toast.error(error.message);
        });
    }

    return (
        <Wrapper>
            <EditEmailChangeModal
                formFieldsBeforeUpdate={formFieldsBeforeUpdate}
                defaultFormFields={defaultFormFields}
                authUser={authUser}
                form={form}
                updateUserInfo={updateUserInfo}
                setShowEmailChangeAlertModal={setShowEmailChangeAlertModal}
                setFormFieldsBeforeUpdate={setFormFieldsBeforeUpdate}
                showEmailChangeAlertModal={showEmailChangeAlertModal}
            />

            <VerifyPhoneModal
                showPhoneVerifyModal={showPhoneVerifyModal}
                sendPhoneVerification={sendPhoneVerification}
                setShowPhoneVerifyModal={setShowPhoneVerifyModal}
                message={phoneVerifyModalMessage}
                setPhoneVerifyModalMessage={setPhoneVerifyModalMessage}
                cancelUpdateProfile={props.cancelUpdateProfile} />

            <SyrfFormWrapper className="no-background">
                <Spin spinning={isUpdatingProfile} tip="Updating your profile...">
                    <Form
                        onValuesChange={() => setFormHasBeenChanged(true)}
                        form={form}
                        layout="vertical"
                        name="basic"
                        initialValues={{
                            email: getUserAttribute(authUser, 'email'),
                            last_name: getUserAttribute(authUser, 'custom:last_name'),
                            first_name: getUserAttribute(authUser, 'custom:first_name'),
                            bio: getUserAttribute(authUser, 'custom:bio'),
                            phone_number: getUserAttribute(authUser, 'phone_number'),
                            address: getUserAttribute(authUser, 'address'),
                            birthdate: moment(getUserAttribute(authUser, 'birthdate')),
                            sailing_number: getUserAttribute(authUser, 'custom:sailing_number'),
                            facebook: getUserAttribute(authUser, 'custom:facebook'),
                            instagram: getUserAttribute(authUser, 'custom:instagram'),
                            twitter: getUserAttribute(authUser, 'custom:twitter'),
                            language: getUserAttribute(authUser, 'custom:language'),
                            country: getUserAttribute(authUser, 'locale'),
                            share_social: !!getUserAttribute(authUser, 'custom:share_social'),
                        }}
                        onFinish={onFinish}
                    >
                        <PublicUserInformation
                            authUser={authUser}
                            cancelUpdateProfile={props.cancelUpdateProfile} />

                        <PrivateUserInformation
                            setShowPhoneVerifyModal={setShowPhoneVerifyModal}
                            address={address} setAddress={setAddress}
                            sendPhoneVerification={sendPhoneVerification}
                            authUser={authUser} />

                        <Form.Item>
                            <StyledSyrfFormButtonWrapper>
                                <SyrfFormButton disabled={!formHasBeenChanged} type="primary" htmlType="submit">
                                    Save
                                </SyrfFormButton>
                            </StyledSyrfFormButtonWrapper>
                        </Form.Item>
                    </Form>
                    <DisclaimerText>* Your personal details will never be shared with 3rd party apps without your permission and will never be sold to advertisers.</DisclaimerText>
                </Spin >
            </SyrfFormWrapper >
        </Wrapper >
    )
}

const Wrapper = styled.div`
    margin-top: 50px;
    width: 100%;
    display:flex;
    justify-content: center;
`;

const DisclaimerText = styled.div`
    font-size: 13px;
    padding: 0 15px;

    ${media.medium`
        padding: 0;
    `}
`;

const StyledSyrfFormButtonWrapper = styled.div`
    padding: 0 15px;

    ${media.medium`
        padding: 0;
    `}
`;