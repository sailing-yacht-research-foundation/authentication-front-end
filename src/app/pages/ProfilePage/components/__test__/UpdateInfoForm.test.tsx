
import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { UpdateInfo } from '../UpdateInfoForm';

const shallowRenderer = createRenderer();

describe('<UpdateInfo />', () => {
  it('should render to match the snapshot', () => {
    shallowRenderer.render(<UpdateInfo />);
    const renderedOutput = shallowRenderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});