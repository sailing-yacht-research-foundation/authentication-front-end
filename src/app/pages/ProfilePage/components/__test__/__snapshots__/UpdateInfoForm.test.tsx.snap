// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateInfo /> should render to match the snapshot 1`] = `
<styled.div>
  <DeleteUserModal
    setShowDeleteUserModal={[Function]}
    showDeleteUserModal={false}
  />
  <styled.div>
    <styled.div>
      <ChangeAvatar />
    </styled.div>
    <Spin
      size="default"
      spinning={false}
      tip="Updating your profile..."
      wrapperClassName=""
    >
      <styled.div>
        Change User Information here
      </styled.div>
      <ForwardRef(InternalForm)
        initialValues={
          Object {
            "address": null,
            "birthdate": null,
            "email": null,
            "facebook": null,
            "instagram": null,
            "language": null,
            "name": null,
            "phone_number": null,
            "sailing_number": null,
            "twitter": null,
          }
        }
        layout="vertical"
        name="basic"
        onFinish={[Function]}
      >
        <Row
          gutter={24}
        >
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label={
                <styled.label>
                  Email
                </styled.label>
              }
              name="email"
              rules={
                Array [
                  Object {
                    "required": true,
                    "type": "email",
                  },
                ]
              }
            >
              <Styled(Input)
                disabled={true}
              />
            </FormItem>
          </Col>
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label={
                <styled.label>
                  Name
                </styled.label>
              }
              name="name"
              rules={
                Array [
                  Object {
                    "required": true,
                  },
                ]
              }
            >
              <Styled(Input) />
            </FormItem>
          </Col>
        </Row>
        <Divider />
        <FormItem
          label={
            <styled.label>
              Address
            </styled.label>
          }
          name="address"
        >
          <PlacesAutocomplete
            debounce={200}
            highlightFirstSuggestion={false}
            onChange={[Function]}
            onError={[Function]}
            searchOptions={Object {}}
            shouldFetchSuggestions={true}
            value=""
          >
            [Function]
          </PlacesAutocomplete>
        </FormItem>
        <Row
          gutter={24}
        >
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label={
                <styled.label>
                  Date Of Birth
                </styled.label>
              }
              name="birthdate"
              rules={
                Array [
                  Object {
                    "required": true,
                    "type": "date",
                  },
                ]
              }
            >
              <Picker
                className="syrf-datepicker"
                dateRender={[Function]}
                disabledDate={[Function]}
                showToday={false}
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
            </FormItem>
          </Col>
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label={
                <styled.label>
                  World Sailing Number
                </styled.label>
              }
              name="sailing_number"
            >
              <Styled(Input) />
            </FormItem>
          </Col>
        </Row>
        <Row
          gutter={24}
        >
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label={
                <styled.label>
                  Phone Number
                </styled.label>
              }
              name="phone_number"
              rules={
                Array [
                  Object {
                    "type": "string",
                  },
                ]
              }
            >
              <Styled(t)
                buttonClass="syrf-flag-dropdown"
                inputClass="syrf-phone-number-input"
                placeholder="Enter phone number"
              />
            </FormItem>
          </Col>
          <Col
            lg={12}
            md={12}
            sm={24}
            xs={24}
          >
            <FormItem
              label="Language"
              name="language"
              rules={
                Array [
                  Object {
                    "required": true,
                  },
                ]
              }
            >
              <Styled(Component)
                filterOption={[Function]}
                placeholder="Select a Language"
                showSearch={true}
              >
                <Option
                  value="zh"
                >
                  中文
                </Option>
                <Option
                  value="nl"
                >
                  Nederlands
                </Option>
                <Option
                  value="en"
                >
                  English
                </Option>
                <Option
                  value="de"
                >
                  Deutsch
                </Option>
                <Option
                  value="es"
                >
                  español
                </Option>
                <Option
                  value="it"
                >
                  Italiano
                </Option>
              </Styled(Component)>
            </FormItem>
          </Col>
        </Row>
        <Divider />
        <FormItem>
          <Styled(Button)
            htmlType="submit"
            type="primary"
          >
            Update Information
          </Styled(Button)>
        </FormItem>
      </ForwardRef(InternalForm)>
    </Spin>
    <styled.div>
      <a
        href="/"
        onClick={[Function]}
      >
        Delete your account
      </a>
    </styled.div>
  </styled.div>
</styled.div>
`;
